### ISSUE TITLE
macCatalyst app with macOS bundle (critical)

### ISSUE THREAD

**Issue by c128128** [tuist/tuist#6798]
### What problem or need do you have?

How to create a target with `destinations: [.macCatalyst]` and dependency a macOS bundle. currently I am getting:

```
Target Example has platforms 'iOS' and product 'application' and depends on target MacBundle of type 'bundle' and platforms 'macOS' which is an invalid or not yet supported combination.
```

This is a very common situation: If you want to call some AppKit APIs, you create a macOS bundle and a protocol, then call them from iOS.

### macOS version

14.6.1

### Tuist version

4.28.1

### Xcode version

16.0.0

**Comment by c128128** [comment]
[Here](https://betterprogramming.pub/how-to-access-the-appkit-api-from-mac-catalyst-apps-2184527020b5) in section B. is an example

**Comment by c128128** [comment]
I've played around with this a bit, and it seems that, at the moment, `tuist` can only generate a predefined dependency graph.

In our project, we have something like this:

<img width="516" alt="Screenshot 2024-10-03 at 14 32 56" src="https://github.com/user-attachments/assets/49b22c6d-bb3a-4705-934a-e6b26ff30577">

Moreover, at the moment, it is impossible to control what we should embed and where. For example:

<img width="506" alt="Screenshot 2024-10-03 at 14 37 04" src="https://github.com/user-attachments/assets/ea99109c-7c5e-49e8-92db-3f90d71b66e3">

**Comment by c128128** [comment]
@fortmarek I believe the current implementation is somewhat inflexible, and you may encounter more issues like this. A better approach might be to allow the end user to define how dependencies should be embedded. To achieve this, the current API would need some adjustments, though I think it’s possible to maintain backward compatibility.

The idea is to let users decide how to embed each dependency in the main target. For example:

```swift
import ProjectDescription

let project = Project(
    name: "Test",
    targets: [
        .target(
            name: "MacBundle",
            destinations: [.mac],
            product: .bundle,
            bundleId: "io.tuist.Test.bundle",
            sources: ["MacBundle/*.swift"]
        ),
        .target(
            name: "Test",
            destinations: [.iPhone, .iPad, .macCatalyst],
            product: .app,
            bundleId: "io.tuist.Test",
            sources: ["Test/Sources/**"],
            resources: ["Test/Resources/**"],
            dependencies: [
                .target(
                    name: "MacBundle",
                    condition: .when([.catalyst]),
                    embed: .init(
                        name: "Embed Bundle",
                        destination: .pluginsAndFoundation,
                        subpath: nil,
                        copyOnlyWhenInstalling: false,
                        filters: [.catalyst],
                        codesignOnCopy: false
                    )
                )
            ]
        )
    ]
)
```

Let me know what you think about this.

So basically, this new `struct` should include all of the following:

<img width="643" alt="Screenshot 2024-10-03 at 17 56 22" src="https://github.com/user-attachments/assets/71ebb9ba-b790-42ec-82c2-1186f510085f">

And in `destinations` this:

<img width="510" alt="Screenshot 2024-10-03 at 17 48 47" src="https://github.com/user-attachments/assets/cf2b3050-9f41-49d0-a1de-a10c2c20658a">

**Comment by fortmarek** [comment]
We don't want to become a 1-1 mapping of _all_ Xcode properties. Wherever we can abstract away some of the Xcode complexity, we try to do so. For example, why would you add `filters: [.catalyst]` when that's already covered by the `TargetDependency.condition`?

As mentioned, if there are things where Tuist doesn't make the right decision, could you describe that usecase with a clear sample and we can take a look? If there are clear cases for where you need fine-grained control, we're opened to that, but our default answer is not to copy everything we see in Xcode.

**Comment by c128128** [comment]
Got it. I will prepare a sample project with a real-life example.

> For example, why would you add filters: [.catalyst] when that's already covered by the TargetDependency.condition?

Agree, that doesn't make sense.
